// @livekit/react-native/lib/commonjs/hooks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  useConnectionState: true,
  useDataChannel: true,
  useIsSpeaking: true,
  useLocalParticipant: true,
  useLocalParticipantPermissions: true,
  useParticipantInfo: true,
  useParticipants: true,
  useRemoteParticipants: true,
  useRemoteParticipant: true,
  useSpeakingParticipants: true,
  useSortedParticipants: true,
  useChat: true,
  useIsEncrypted: true,
  useRoomInfo: true,
  useIsMuted: true,
  useParticipantTracks: true,
  useLiveKitRoom: true,
  RoomContext: true,
  useRoomContext: true,
  ParticipantContext: true,
  useParticipantContext: true,
  TrackRefContext: true,
  useTrackRefContext: true,
  useTracks: true,
  isTrackReference: true,
  useEnsureTrackRef: true,
  useMaybeTrackRefContext: true,
  useTrackMutedIndicator: true,
  useVisualStableUpdate: true,
  useVoiceAssistant: true,
  useTrackTranscription: true
};
Object.defineProperty(exports, "ParticipantContext", {
  enumerable: true,
  get: function () {
    return _componentsReact.ParticipantContext;
  }
});
Object.defineProperty(exports, "RoomContext", {
  enumerable: true,
  get: function () {
    return _componentsReact.RoomContext;
  }
});
Object.defineProperty(exports, "TrackRefContext", {
  enumerable: true,
  get: function () {
    return _componentsReact.TrackRefContext;
  }
});
Object.defineProperty(exports, "isTrackReference", {
  enumerable: true,
  get: function () {
    return _componentsReact.isTrackReference;
  }
});
Object.defineProperty(exports, "useChat", {
  enumerable: true,
  get: function () {
    return _componentsReact.useChat;
  }
});
Object.defineProperty(exports, "useConnectionState", {
  enumerable: true,
  get: function () {
    return _componentsReact.useConnectionState;
  }
});
Object.defineProperty(exports, "useDataChannel", {
  enumerable: true,
  get: function () {
    return _componentsReact.useDataChannel;
  }
});
Object.defineProperty(exports, "useEnsureTrackRef", {
  enumerable: true,
  get: function () {
    return _componentsReact.useEnsureTrackRef;
  }
});
Object.defineProperty(exports, "useIsEncrypted", {
  enumerable: true,
  get: function () {
    return _componentsReact.useIsEncrypted;
  }
});
Object.defineProperty(exports, "useIsMuted", {
  enumerable: true,
  get: function () {
    return _componentsReact.useIsMuted;
  }
});
Object.defineProperty(exports, "useIsSpeaking", {
  enumerable: true,
  get: function () {
    return _componentsReact.useIsSpeaking;
  }
});
Object.defineProperty(exports, "useLiveKitRoom", {
  enumerable: true,
  get: function () {
    return _componentsReact.useLiveKitRoom;
  }
});
Object.defineProperty(exports, "useLocalParticipant", {
  enumerable: true,
  get: function () {
    return _componentsReact.useLocalParticipant;
  }
});
Object.defineProperty(exports, "useLocalParticipantPermissions", {
  enumerable: true,
  get: function () {
    return _componentsReact.useLocalParticipantPermissions;
  }
});
Object.defineProperty(exports, "useMaybeTrackRefContext", {
  enumerable: true,
  get: function () {
    return _componentsReact.useMaybeTrackRefContext;
  }
});
Object.defineProperty(exports, "useParticipantContext", {
  enumerable: true,
  get: function () {
    return _componentsReact.useParticipantContext;
  }
});
Object.defineProperty(exports, "useParticipantInfo", {
  enumerable: true,
  get: function () {
    return _componentsReact.useParticipantInfo;
  }
});
Object.defineProperty(exports, "useParticipantTracks", {
  enumerable: true,
  get: function () {
    return _componentsReact.useParticipantTracks;
  }
});
Object.defineProperty(exports, "useParticipants", {
  enumerable: true,
  get: function () {
    return _componentsReact.useParticipants;
  }
});
Object.defineProperty(exports, "useRemoteParticipant", {
  enumerable: true,
  get: function () {
    return _componentsReact.useRemoteParticipant;
  }
});
Object.defineProperty(exports, "useRemoteParticipants", {
  enumerable: true,
  get: function () {
    return _componentsReact.useRemoteParticipants;
  }
});
Object.defineProperty(exports, "useRoomContext", {
  enumerable: true,
  get: function () {
    return _componentsReact.useRoomContext;
  }
});
Object.defineProperty(exports, "useRoomInfo", {
  enumerable: true,
  get: function () {
    return _componentsReact.useRoomInfo;
  }
});
Object.defineProperty(exports, "useSortedParticipants", {
  enumerable: true,
  get: function () {
    return _componentsReact.useSortedParticipants;
  }
});
Object.defineProperty(exports, "useSpeakingParticipants", {
  enumerable: true,
  get: function () {
    return _componentsReact.useSpeakingParticipants;
  }
});
Object.defineProperty(exports, "useTrackMutedIndicator", {
  enumerable: true,
  get: function () {
    return _componentsReact.useTrackMutedIndicator;
  }
});
Object.defineProperty(exports, "useTrackRefContext", {
  enumerable: true,
  get: function () {
    return _componentsReact.useTrackRefContext;
  }
});
Object.defineProperty(exports, "useTrackTranscription", {
  enumerable: true,
  get: function () {
    return _componentsReact.useTrackTranscription;
  }
});
Object.defineProperty(exports, "useTracks", {
  enumerable: true,
  get: function () {
    return _componentsReact.useTracks;
  }
});
Object.defineProperty(exports, "useVisualStableUpdate", {
  enumerable: true,
  get: function () {
    return _componentsReact.useVisualStableUpdate;
  }
});
Object.defineProperty(exports, "useVoiceAssistant", {
  enumerable: true,
  get: function () {
    return _componentsReact.useVoiceAssistant;
  }
});
var _componentsReact = require("@livekit/components-react");
var _useE2EEManager = require("./hooks/useE2EEManager");
Object.keys(_useE2EEManager).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useE2EEManager[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useE2EEManager[key];
    }
  });
});
var _useTrackVolume = require("./hooks/useTrackVolume");
Object.keys(_useTrackVolume).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useTrackVolume[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useTrackVolume[key];
    }
  });
});
var _useMultibandTrackVolume = require("./hooks/useMultibandTrackVolume");
Object.keys(_useMultibandTrackVolume).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useMultibandTrackVolume[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useMultibandTrackVolume[key];
    }
  });
});
