// @livekit/react-native/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  registerGlobals: true,
  AudioSession: true,
  AndroidAudioTypePresets: true,
  getDefaultAppleAudioConfigurationForMode: true,
  RNE2EEManager: true,
  RNKeyProvider: true
};
Object.defineProperty(exports, "AndroidAudioTypePresets", {
  enumerable: true,
  get: function () {
    return _AudioSession.AndroidAudioTypePresets;
  }
});
Object.defineProperty(exports, "AudioSession", {
  enumerable: true,
  get: function () {
    return _AudioSession.default;
  }
});
Object.defineProperty(exports, "RNE2EEManager", {
  enumerable: true,
  get: function () {
    return _RNE2EEManager.default;
  }
});
Object.defineProperty(exports, "RNKeyProvider", {
  enumerable: true,
  get: function () {
    return _RNKeyProvider.default;
  }
});
Object.defineProperty(exports, "getDefaultAppleAudioConfigurationForMode", {
  enumerable: true,
  get: function () {
    return _AudioSession.getDefaultAppleAudioConfigurationForMode;
  }
});
exports.registerGlobals = registerGlobals;
require("well-known-symbols/Symbol.asyncIterator/auto");
require("well-known-symbols/Symbol.iterator/auto");
require("./polyfills/MediaRecorderShim");
require("react-native-quick-base64");
var _reactNativeWebrtc = require("@livekit/react-native-webrtc");
var _reactNativeUrlPolyfill = require("react-native-url-polyfill");
require("./polyfills/EncoderDecoderTogether.min.js");
var _AudioSession = _interopRequireWildcard(require("./audio/AudioSession"));
var _reactNative = require("react-native");
var _RNE2EEManager = _interopRequireDefault(require("./e2ee/RNE2EEManager"));
var _RNKeyProvider = _interopRequireDefault(require("./e2ee/RNKeyProvider"));
var _EventEmitter = require("./events/EventEmitter");
var _webStreamsPolyfill = require("web-streams-polyfill");
var _hooks = require("./hooks");
Object.keys(_hooks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _hooks[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _hooks[key];
    }
  });
});
var _BarVisualizer = require("./components/BarVisualizer");
Object.keys(_BarVisualizer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _BarVisualizer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _BarVisualizer[key];
    }
  });
});
var _LiveKitRoom = require("./components/LiveKitRoom");
Object.keys(_LiveKitRoom).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _LiveKitRoom[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _LiveKitRoom[key];
    }
  });
});
var _VideoTrack = require("./components/VideoTrack");
Object.keys(_VideoTrack).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _VideoTrack[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _VideoTrack[key];
    }
  });
});
var _VideoView = require("./components/VideoView");
Object.keys(_VideoView).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _VideoView[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _VideoView[key];
    }
  });
});
var _useParticipant = require("./useParticipant");
Object.keys(_useParticipant).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useParticipant[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useParticipant[key];
    }
  });
});
var _useRoom = require("./useRoom");
Object.keys(_useRoom).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useRoom[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useRoom[key];
    }
  });
});
var _logger = require("./logger");
Object.keys(_logger).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _logger[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _logger[key];
    }
  });
});
var _AudioManager = require("./audio/AudioManager");
Object.keys(_AudioManager).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _AudioManager[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _AudioManager[key];
    }
  });
});
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * Registers the required globals needed for LiveKit to work.
 *
 * Must be called before using LiveKit.
 */
function registerGlobals() {
  (0, _reactNativeWebrtc.registerGlobals)();
  iosCategoryEnforce();
  livekitRegisterGlobals();
  (0, _reactNativeUrlPolyfill.setupURLPolyfill)();
  fixWebrtcAdapter();
  shimPromiseAllSettled();
  shimArrayAt();
  shimCryptoUuid();
  shimWebstreams();
  (0, _EventEmitter.setupNativeEvents)();
}

/**
 * Enforces changing to playAndRecord category prior to obtaining microphone.
 */
function iosCategoryEnforce() {
  if (_reactNative.Platform.OS === 'ios') {
    // @ts-ignore
    let getUserMediaFunc = global.navigator.mediaDevices.getUserMedia;
    // @ts-ignore
    global.navigator.mediaDevices.getUserMedia = async constraints => {
      if (constraints.audio) {
        await _AudioSession.default.setAppleAudioConfiguration({
          audioCategory: 'playAndRecord'
        });
      }
      return await getUserMediaFunc(constraints);
    };
  }
}
function livekitRegisterGlobals() {
  let lkGlobal = {
    platform: _reactNative.Platform.OS,
    devicePixelRatio: _reactNative.PixelRatio.get()
  };

  // @ts-ignore
  global.LiveKitReactNativeGlobal = lkGlobal;
}
function fixWebrtcAdapter() {
  var _window;
  // @ts-ignore
  if (((_window = window) === null || _window === void 0 ? void 0 : _window.navigator) !== undefined) {
    // @ts-ignore
    const {
      navigator
    } = window;
    if (navigator.userAgent === undefined) {
      navigator.userAgent = navigator.product ?? 'Unknown';
    }
  }
}
function shimPromiseAllSettled() {
  var allSettled = require('promise.allsettled');
  allSettled.shim();
}
function shimArrayAt() {
  // Some versions of RN don't have Array.prototype.at, which is used by sdp-transform
  if (!Array.prototype.at) {
    var at = require('array.prototype.at');
    at.shim();
  }
}
function shimCryptoUuid() {
  var _global$crypto;
  let crypto = global.crypto;
  if (typeof ((_global$crypto = global.crypto) === null || _global$crypto === void 0 ? void 0 : _global$crypto.randomUUID) !== 'function') {
    let createRandomUUID = () => {
      return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
        /* eslint-disable no-bitwise */
        const r = Math.random() * 16 | 0;
        const v = c === 'x' ? r : r & 0x3 | 0x8;
        return v.toString(16);
      });
    };
    if (!crypto) {
      crypto = {};
      global.crypto = crypto;
    }
    crypto.randomUUID = createRandomUUID;
  }
}
function shimWebstreams() {
  // @ts-expect-error: global.WritableStream isn't typed here.
  if (typeof global.WritableStream === 'undefined') {
    // @ts-expect-error
    global.WritableStream = _webStreamsPolyfill.WritableStream;
  }

  // @ts-expect-error: global.ReadableStream isn't typed here.
  if (typeof global.ReadableStream === 'undefined') {
    // @ts-expect-error
    global.ReadableStream = _webStreamsPolyfill.ReadableStream;
  }
}

// deprecated
// deprecated
// deprecated
